### PR TITLE
Test for Neumann BC under PETSc and unified known solution applying interface.

### PR DESCRIPTION
**Pull request by wenqing**
1. Unified known solution applying interface for Dirichlet BC.
2. Corrected index type in all that is used in the equation assembly.
3. Added a Neumann type test for FEM#PETSc.

### CODE REVIEW

File: MathLib/LinAlg/Eigen/EigenTools.h [hunk 1]
```diff
@@ -11,12 +11,14 @@
...
 
 #include <vector>
+#include <Eigen/Sparse> //for RawMatrixType::Index
 
 namespace MathLib
 {
 class EigenMatrix;
 class EigenVector;
 
+using RawMatrixType = Eigen::SparseMatrix<double, Eigen::RowMajor>;
```
Comment (chleh) [inline MathLib/LinAlg/Eigen/EigenTools.h]: Isn't that already defined inside the `EigenMatrix` class? I would prefer to reuse that definition.

File: MathLib/LinAlg/Dense/DenseTools.h [hunk 2]
```diff
@@ -16,12 +16,11 @@
...
 
 #include <vector>
-#include "DenseMatrix.h"
-#include "DenseVector.h"
+#include "GlobalDenseMatrix.h"
 
 namespace MathLib
 {
-
+using DenseMatrixIndexType = GlobalDenseMatrix<double>::IndexType;
```
Comment (chleh) [inline MathLib/LinAlg/Dense/DenseTools.h]: I would prefer to drop that typedef.

File: MathLib/LinAlg/Dense/DenseTools.cpp [hunk 3]
```diff
@@ -17,8 +17,8 @@
 namespace MathLib
 {
 
-void applyKnownSolution(DenseMatrix<double> &eqsA, DenseVector<double> &eqsRHS, std::size_t k,
-		double val)
+void applyKnownSolution(DenseMatrix<double> &eqsA, DenseVector<double> &eqsRHS,
+		DenseMatrixIndexType k, double val)
```
Comment (chleh) [inline MathLib/LinAlg/Dense/DenseTools.cpp]: Strictly speaking, this function deals with type `DenseMatrix`, so the index type should be `DenseMatrix<double>::IdxType` and not a typedef from `GlobalDenseMatrix`.

File: MathLib/LinAlg/Lis/LisTools.h [hunk 4]
```diff
@@ -34,8 +35,8 @@ class LisVector;
  * @param rows a vector of known solution entry IDs
  * @param vals a vector of known solutions
  */
-void applyKnownSolution(LisMatrix &eqsA, LisVector &eqsRHS,
-	const std::vector<std::size_t> &rows,
+void applyKnownSolution(LisMatrix &eqsA, LisVector &eqsRHS, LisVector &/*eqsX*/,
+	const std::vector< LIS_INT> &rows, //LIS_INT has an alias of LisMatrix::IndexType
```
Comment (chleh) [inline MathLib/LinAlg/Lis/LisTools.h]: Completely consistent with LisTools/EigenTools would be to use the mentioned alias and not LIS_INT.

File: ProcessLib/GroundwaterFlowProcess.h [hunk 5]
```diff
@@ -289,33 +289,12 @@ class GroundwaterFlowProcess : public Process
         // Call global assembler for each local assembly item.
         _global_setup.execute(*_global_assembler, _local_assemblers);
 
-#ifdef USE_PETSC
```
Comment (chleh) [inline ProcessLib/GroundwaterFlowProcess.h]: Great improvement that now we don't need to make that distinction anymore!

File: ProcessLib/UniformDirichletBoundaryCondition.h [hunk 6]
```diff
@@ -67,16 +69,13 @@ class UniformDirichletBoundaryCondition
                                 MeshLib::MeshItemType::Node,
                                 id);
             // TODO: that might be slow, but only done once
-            id = dof_table.getGlobalIndex(l, component_id);
+            const auto g_idx = dof_table.getGlobalIndex(l, component_id);
+            if (g_idx >= 0)
+            {
+                global_ids.emplace_back(g_idx);
```
Comment (chleh) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: Maybe it's worth to call `reserve()` on global_ids before the loop.

File: ProcessLib/UniformDirichletBoundaryCondition.h [hunk 7]
```diff
@@ -67,16 +69,13 @@ class UniformDirichletBoundaryCondition
                                 MeshLib::MeshItemType::Node,
                                 id);
             // TODO: that might be slow, but only done once
-            id = dof_table.getGlobalIndex(l, component_id);
+            const auto g_idx = dof_table.getGlobalIndex(l, component_id);
+            if (g_idx >= 0)
```
Comment (chleh) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: Maybe it there should be a comment that this is Petsc-specific behaviour?

File: Tests/AssemblerLib/SteadyDiffusion2DExample1.h [hunk 8]
```diff
@@ -140,7 +140,7 @@ struct SteadyDiffusion2DExample1
 	static const std::size_t mesh_stride = mesh_subdivs + 1;
 	static const std::size_t dim_eqs = mesh_stride * mesh_stride;
 	MeshLib::Mesh* msh;
-	std::vector<std::size_t> vec_DirichletBC_id;
+	std::vector<int> vec_DirichletBC_id;
```
Comment (chleh) [inline Tests/AssemblerLib/SteadyDiffusion2DExample1.h]: Should it be some `IndexType`? But in my opinion, it's a test. As long as it works, it's fine. If will fail due to the index type, we can easily fix it later on.

**Comment by wenqing** [comment]
@endJunction  Compilation errors under LIS have been removed.

File: Tests/MathLib/TestLinearSolver.cpp [hunk 9]
```diff
@@ -207,10 +207,12 @@ TEST(MathLib, CheckInterface_GaussAlgorithm)
     boost::property_tree::ptree t_solver;
     t_root.put_child("ogs", t_solver);
 
+    using Example = Example1<std::size_t>; 
+
     typedef MathLib::GaussAlgorithm<MathLib::GlobalDenseMatrix<double>, MathLib::DenseVector<double> > LinearSolverType;
-    MathLib::GlobalDenseMatrix<double> A(Example1::dim_eqs, Example1::dim_eqs);
+    MathLib::GlobalDenseMatrix<double> A(Example::dim_eqs, Example::dim_eqs);
```
Comment (chleh) [inline Tests/MathLib/TestLinearSolver.cpp]: Why the change `Example1` → `Example`? I don't see it immediately.

**Comment by chleh** [comment]
I only saw small issues that are very easy to address.

File: MathLib/LinAlg/Eigen/EigenTools.h (hunk 1 above)
Comment (wenqing) [inline MathLib/LinAlg/Eigen/EigenTools.h]: Just want to take the advantage of the forward class declaration, and to avoid include the header of the matrix classes that have the definition of IndexType.

File: MathLib/LinAlg/Dense/DenseTools.h (hunk 2 above)
Comment (wenqing) [inline MathLib/LinAlg/Dense/DenseTools.h]: Same reason as before.

File: MathLib/LinAlg/Dense/DenseTools.cpp (hunk 3 above)
Comment (wenqing) [inline MathLib/LinAlg/Dense/DenseTools.cpp]: Will be changed.

File: MathLib/LinAlg/Lis/LisTools.h (hunk 4 above)
Comment (wenqing) [inline MathLib/LinAlg/Lis/LisTools.h]: Same reason as above: just to avoid include LisMatrix.h

File: ProcessLib/UniformDirichletBoundaryCondition.h (hunk 6 above)
Comment (wenqing) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: The thing is that we have to count how many non-ghost entries in advance in order to call reserve().

File: ProcessLib/UniformDirichletBoundaryCondition.h (hunk 7 above)
Comment (wenqing) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: Will add a comment.

File: ProcessLib/UniformDirichletBoundaryCondition.h (hunk 6 above)
Comment (chleh) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: Possibly it would be slightly more performant to reserve space for all entries of `ids`. Sure, that would waste some space, but memory consumption won't blow up by that.
But I have not measured the performace impact and this function is only called once per simulation (or in future maby once per timestep).
So in my opinion: It would be nice if you kept the `global_ids.reserve(global_ids.size() + ids.size());`, at least if you do not strongly object against the waste of space.

Comment (wenqing) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: Take your idea.

File: ProcessLib/UniformDirichletBoundaryCondition.h (hunk 7 above)
Comment (wenqing) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: Added.

File: Tests/AssemblerLib/SteadyDiffusion2DExample1.h (hunk 8 above)
Comment (wenqing) [inline Tests/AssemblerLib/SteadyDiffusion2DExample1.h]: Changed by turning class SteadyDiffusion2DExample1.h into a template class.

File: Tests/MathLib/TestLinearSolver.cpp (hunk 9 above)
Comment (wenqing) [inline Tests/MathLib/TestLinearSolver.cpp]: Above line 210, there is a newly introduced alias, Example, to ``Example1<std::size_t>`` as
```
using Example = Example1<std::size_t>; 
```

**Comment by wenqing** [comment]
@chleh  Your comments have been addressed (See commit d487ad0).

File: Tests/MathLib/TestLinearSolver.cpp (hunk 9 above)
Comment (chleh) [inline Tests/MathLib/TestLinearSolver.cpp]: Thanks.

File: ProcessLib/UniformDirichletBoundaryCondition.h [hunk 10]
```diff
@@ -61,22 +63,27 @@ class UniformDirichletBoundaryCondition
...
+            const auto g_idx = dof_table.getGlobalIndex(l, component_id);
+            // For the DDC approach (e.g. with PETSc option), the negative
+            // index of g_idx means that the entry by that index is a ghost one,
+            // which should be dropped. Especially for PETSc routines MatZeroRows
+            // and MatZeroRowsColumns, which are called to apply the Dirichlet BC,
+            // the negative index is not accepted like other matrix or vector
+            // PETSc routines. Therefore, the following if-condition is applied.  
+            if (g_idx >= 0)
+            {
+                global_ids.push_back(g_idx);
```
Comment (chleh) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: Just a short note: You could have kept it as emplace_back.

**Comment by chleh** [comment]
:+1:

File: ProcessLib/UniformDirichletBoundaryCondition.h (hunk 10 above)
Comment (wenqing) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: Reverted to  emplace_back by amending (see the amended commit 68f93c1).

File: MathLib/LinAlg/Dense/DenseTools.cpp (hunk 3 above)
Comment (wenqing) [inline MathLib/LinAlg/Dense/DenseTools.cpp]: Changed.

File: SimpleTests/MeshTests/MPI/CMakeLists.txt [hunk 11]
```diff
@@ -43,10 +44,16 @@ ExternalData_Add_Test(
...
-set(FilePath "DATA{${ExternalData_SOURCE_ROOT}/EllipticPETSc/,REGEX:.*}")
 set(MPITestParameters -np 3 "${PROJECT_BINARY_DIR}/bin/ogs" "${FilePath}/cube_1e3.prj" --  -gw_ksp_type bcgs -gw_pc_type bjacobi -gw_ksp_atol 1.e-10)
 ExternalData_Add_Test(
     data
-    NAME ParallelFEM_GroundWaterFlow3D
+    NAME ParallelFEM_GroundWaterFlow3D_DirichletBC
+    COMMAND  "mpirun" ${MPITestParameters} ${SolverType}
+)
+
+set(MPITestParameters -np 3 "${PROJECT_BINARY_DIR}/bin/ogs" "${FilePath}/cube_1e3_neumann.prj" --  -gw_ksp_type bcgs -gw_pc_type bjacobi -gw_ksp_atol 1.e-10)
```
Comment (endJunction) [inline SimpleTests/MeshTests/MPI/CMakeLists.txt]: For another PR I suggest to put the linear solver options into the project file; this way the simulation will run independent of the way it was started---by `ctest` or manually from prompt.
